File: dynamic_listings/lua/default/collisiondetection/line_rectangle.lua
local function lineLineCollision(x1, y1, x2, y2, x3, y3, x4, y4)
    -- our previous implementation of the line/line collision detection
    -- ...
end

local function pointRectCollision(x1, y1, rectx, recty, rectwidth, rectheight)
    -- our previous implementation of a point/rectangle collision detection
    -- ...
end

local function lineRectangleCollision(x1, y1, x2, y2, rectx, recty, rectwidth, rectheight)
    -- If we want to test if a line is completely inside of a rect, we just need
    -- to see if any of its endpoints is inside the rectangle
    if (pointRectCollision(x1, y1, rectx, recty, rectwidth, rectheight) or pointRectCollision(x2, y2, rectx, recty, rectwidth, rectheight)) then
        -- At least one of the ends of the segment is inside the rectangle
        return true
    end
    -- Now to test the rectangle against the line, if it's not completely inside
    local left = lineLineCollision(x1, y1, x2, y2, rectx, recty, rectx, recty + rectheight)
    local right = lineLineCollision(x1, y1, x2, y2, rectx + rectwidth, recty, rectx + rectwidth, recty + rectheight)
    local top = lineLineCollision(x1, y1, x2, y2, rectx, recty, rectx + rectwidth, recty)
    local bottom = lineLineCollision(x1, y1, x2, y2, rectx, recty + rectheight, rectx + rectwidth, recty + rectheight)

    if (left or right or top or bottom) then
        -- We hit one of the sides, we are colliding
        return true
    end

    -- In any other case, return false
    return false
end

File: dynamic_listings/rust/default/collisiondetection/line_rectangle.rs
# #![allow(dead_code, unused)]
fn line_line_collision(
    x1: f32,
    y1: f32,
    x2: f32,
    y2: f32,
    x3: f32,
    y3: f32,
    x4: f32,
    y4: f32,
) -> bool {
    // our previous implementation of the line/line collision detection
# todo!()
}

fn point_rect_collision(
    x1: f32,
    y1: f32,
    rectx: f32,
    recty: f32,
    rectwidth: f32,
    rectheight: f32,
) -> bool {
    // our previous implementation of a point/rectangle collision detection
# todo!()
}

fn line_rectangle_collision(
    x1: f32,
    y1: f32,
    x2: f32,
    y2: f32,
    rectx: f32,
    recty: f32,
    rectwidth: f32,
    rectheight: f32,
) -> bool {
    // If we want to test if a line is completely inside of a rect, we just need
    // to see if any of its endpoints is inside the rectangle
    if point_rect_collision(x1, y1, rectx, recty, rectwidth, rectheight)
        || point_rect_collision(x2, y2, rectx, recty, rectwidth, rectheight)
    {
        // At least one of the ends of the segment is inside the rectangle
        return true;
    }

    // Now to test the rectangle against the line, if it's not completely inside
    let left = line_line_collision(x1, y1, x2, y2, rectx, recty, rectx, recty + rectheight);
    let right = line_line_collision(
        x1,
        y1,
        x2,
        y2,
        rectx + rectwidth,
        recty,
        rectx + rectwidth,
        recty + rectheight,
    );
    let top = line_line_collision(x1, y1, x2, y2, rectx, recty, rectx + rectwidth, recty);
    let bottom = line_line_collision(
        x1,
        y1,
        x2,
        y2,
        rectx,
        recty + rectheight,
        rectx + rectwidth,
        recty + rectheight,
    );

    if left || right || top || bottom {
        // We hit one of the sides, we are colliding
        return true;
    }

    // In any other case, return false
    return false;
}
